basic/src/test/java/org/hibernate/benchmark/enhancement/AccessOptimizers.java

```diff
@@ -0,0 +1,566 @@
+
+	// WHen true, we make JIT compile all Morphism versions of the method but then run only 1 type
+	@Param({ "false", "true" })
+	private boolean polluteAtWarmup;
+
+	@Param({ "100" })
+	private int count;
+
+	private SessionFactory sessionFactory;
+	private Session session;
+
+	int nextEntityTypeId;
+	int[] entityTypes;
+
+	@Setup
+	public void setup(Blackhole bh) {
+		int types = morphism.types();
+		if ( types == 1 && polluteAtWarmup ) {
+			// this is fairly useless
+			throw new IllegalStateException( "Cannot pollute with monomorphic types" );
+		}
+		switch ( access ) {
+			case OPTIMIZED -> sessionFactory = getSessionFactory( new TestBytecodeProvider(), types, access );
+			case STANDARD -> sessionFactory = getSessionFactory( new ProxyOnlyBytecodeProvider(), types, access );
+		}
+		populateData( sessionFactory, count, types );
+		session = sessionFactory.openSession();
+		session.getTransaction().begin();
+
+		nextEntityTypeId = 0;
+		entityTypes = new int[types];
+		for ( int i = 0; i < types; i++ ) {
+			entityTypes[i] = i;
+		}
+		if ( polluteAtWarmup ) {
+			// this is ensuring that every each queryX method is called enough to be fully compiled
+			for ( int i = 0; i < 10_000; i++ ) {
+				switch ( types ) {
+					case 4:
+						query( EntityOfBasics.class, session, bh );
+					case 3:
+						query( AnotherEntity.class, session, bh );
+					case 2:
+						query( EntityWithLazyOneToOne.class, session, bh );
+					case 1:
+						query( SimpleEntity.class, session, bh );
+				}
+			}
+			// force to make it monomorphic now
+			entityTypes = new int[] { 0 };
+			nextEntityTypeId = 0;
+		}
+	}
+
+	private int nextEntityTypeId() {
+		var entityTypes = this.entityTypes;
+		if ( nextEntityTypeId >= entityTypes.length ) {
+			nextEntityTypeId = 1;
+			return entityTypes[0];
+		}
+		else {
+			return entityTypes[nextEntityTypeId++];
+		}
+	}
+
+	private static void populateData(SessionFactory sf, int count, int types) {
+		if ( types < 1 || types > 4 ) {
+			throw new IllegalArgumentException( "Invalid types" );
+		}
+		final Session session = sf.openSession();
+		session.getTransaction().begin();
+		for ( int i = 1; i <= count; i++ ) {
+			final SimpleEntity simpleEntity = randomSimpleEntity( i );
+			switch ( types ) {
+				case 4:
+					session.persist( randomEntityOfBasics( i ) );
+				case 3:
+					session.persist( randomAnotherEntity( i ) );
+				case 2:
+					session.persist( randomEntityWithToOne( i, simpleEntity ) );
+				case 1:
+					session.persist( simpleEntity );
+			}
+		}
+		session.getTransaction().commit();
+		session.close();
+	}
+
+	private static SimpleEntity randomSimpleEntity(int count) {
+		final long timestamp = System.currentTimeMillis() + count * 3_600_000L;
+		return new SimpleEntity(
+				count,
+				new Date( timestamp ),
+				Instant.ofEpochMilli( timestamp ),
+				count,
+				timestamp,
+				"simple_entity_" + count
+		);
+	}
+
+	private static AnotherEntity randomAnotherEntity(int count) {
```

Comment:
Using basic properties only might not be the most realistic use-case, but identity hascode is used to keep track of persistent collections in the persistence context and it impacts the benchmark greatly, so I went the simple route.